Sources/DocCDocumentation/DocCReferenceResolutionService.swift

```diff
@@ -0,0 +1,234 @@
+    guard let convertRequestIdentifier = request.convertRequestIdentifier else {
+      throw .missingConvertRequestIdentifier
+    }
+    guard let context = context(forKey: convertRequestIdentifier) else {
+      throw .missingContext
+    }
+    switch request.payload {
+    case .symbol(let symbolUSR):
+      throw .symbolNotFound(symbolUSR)
+    case .asset(let assetReference):
+      guard let catalog = context.catalogIndex else {
+        throw .indexNotAvailable
+      }
+      guard let dataAsset = catalog.assets[assetReference.assetName] else {
+        throw .assetNotFound
+      }
+      return .asset(dataAsset)
+    case .topic(let topicURL):
+      // Check if this is a link to another documentation article
+      let relevantPathComponents = topicURL.pathComponents.filter { $0 != "/" }
+      let resolvedReference: TopicRenderReference? =
+        switch relevantPathComponents.first {
+        case NodeURLGenerator.Path.documentationFolderName:
+          context.catalogIndex?.articles[topicURL.lastPathComponent]
+        case NodeURLGenerator.Path.tutorialsFolderName:
+          context.catalogIndex?.tutorials[topicURL.lastPathComponent]
+        default:
+          nil
+        }
+      if let resolvedReference {
+        return .resolvedInformation(OutOfProcessReferenceResolver.ResolvedInformation(resolvedReference, url: topicURL))
+      }
+      // Otherwise this must be a link to a symbol
+      let urlString = topicURL.absoluteString
+      guard let doccSymbolLink = DocCSymbolLink(linkString: urlString) else {
+        throw .invalidURLInRequest
+      }
+      // Don't bother checking to see if the symbol actually exists in the index. This can be time consuming and
+      // it would be better to report errors/warnings for unresolved symbols directly within the document, anyway.
+      return .resolvedInformation(
+        OutOfProcessReferenceResolver.ResolvedInformation(
+          symbolURL: topicURL,
+          symbolName: doccSymbolLink.symbolName
+        )
+      )
+    }
+  }
+
+  private func decode<T>(_ type: T.Type, from data: Data) throws(ReferenceResolutionError) -> T where T: Decodable {
+    do {
+      return try JSONDecoder().decode(T.self, from: data)
+    } catch {
+      throw .decodingFailure(error.localizedDescription)
+    }
+  }
+
+  private func encode<T: Encodable>(_ value: T) throws(ReferenceResolutionError) -> Data {
+    do {
+      return try JSONEncoder().encode(value)
+    } catch {
+      throw .decodingFailure(error.localizedDescription)
+    }
+  }
+
+  private func createResponseWithErrorMessage(_ message: String) -> DocumentationServer.Message {
+    let errorMessage = OutOfProcessReferenceResolver.Response.errorMessage(message)
+    do {
+      let encodedErrorMessage = try JSONEncoder().encode(errorMessage)
+      return createResponse(payload: encodedErrorMessage)
+    } catch {
+      return createResponse(payload: nil)
```

Comment:
Should we log the error here? Just so we’ve got something to indicate a failure should we hit this case.